### PR TITLE
Do major refactor of database code

### PR DESCRIPTION
**Pull request by Others**
This should make it less racy, and more consistent. Also easier
    to read and understand. Works with my v9_database PR. Follow ups
    should switch us over to the "starting paused" system

### CODE REVIEW

**Comment by Others** [comment]
@hjaensch7 can you test this with me tomorrow?

File: database/database.go [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+	return &Driver{
+		db: db,
+	}, nil
+}
+
+// TODO: Consider using transactions throughout here
+
+func (driver *Driver) FindUserID(githubUsername string) (string, error) {
+	var userID string
+	upsertQuery := `INSERT INTO v9.public.users(email, github_username) VALUES (NULL, $1)
```
Comment (pcodes) [inline database/database.go:32]: Add documentation for why you're replacing the github username with...the same username

File: database/database.go [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+}
+
+func (driver *Driver) FindComponentID(compID *worker.ComponentID) (string, error) {
+	userID, err := driver.FindUserID(compID.User)
+	if err != nil {
+		return "", err
+	}
+
+	var compDBID string
+	upsertQuery := `INSERT INTO v9.public.components(user_id, github_repo, deployment_intention) VALUES ($1, $2, $3)
```
Comment (pcodes) [inline database/database.go:52]: Is this the same as above?

File: database/database.go [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+	if compLog.Log != nil {
+		logText = compLog.Log
+	}
+
+	if compLog.Error != nil {
+		logError = compLog.Error
+	}
+
+	insertUpdateQuery := `INSERT INTO v9.public.logs
```
Comment (pcodes) [inline database/database.go]: Naming inconstancy- either stick with upsert or remove the portmanteau

File: database/database.go [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+func (driver *Driver) EnterDeploymentEntry(compID *worker.ComponentID) error {
+	compDBID, err := driver.FindComponentID(compID)
+	if err != nil {
+		return err
+	}
+
+	var deploymentStartTime string
+	upsertQuery := `INSERT INTO v9.public.deploying(component_id, deployment_reason)
+	VALUES ($1, $2) ON CONFLICT DO NOTHING
+	RETURNING to_char(deployment_start_time::timestamp at time zone 'UTC', 'YYYY-MM-DD"T"HH24:MI:SS"Z"')`
```
Comment (pcodes) [inline database/database.go:163]: How does this correspond to TIMESTAMPTZ? It looks like you're using UTC, will this cause issues?

File: pushHandler.go [hunk 5]
```diff
@@ -59,54 +63,69 @@ func (h *pushHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		user = parsedPayload.Repository.Owner.Login
 		repo = parsedPayload.Repository.Name
 	}
-	compID := componentID{user, repo, "test_hash"}
+
+	compID := worker.ComponentID{User: user, Repo: repo, Hash: "test_hash"}
```
Comment (pcodes) [inline pushHandler.go:68]: Can we get a comment that points to the issue? I didn't realize we had one

**Review by pcodes — CHANGES_REQUESTED** [review]
My comments are just about the DB related stuff, @hjaensch7 should also look at the refactoring code. There's some misc issues I've pointed out that should be fixed.

File: database/database.go [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+	if err != nil {
+		return fmt.Errorf("error marshaling latency percentiles: %w", err)
+	}
+
+	insertQuery := `INSERT INTO v9.public.stats
+    (worker_id, component_id, color, stat_window_seconds, hits,
+     avg_response_bytes, avg_ms_latency, ms_latency_percentiles)
+    VALUES ($1, $2, $3, $4, $5, $6, $7, $8)`
+
+	_, err = driver.db.Exec(
```
Comment (hjaensch7) [inline database/database.go:94]: what return isn't being handled here?

File: database/database.go [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+	if compLog.Error != nil {
+		logError = compLog.Error
+	}
+
+	insertUpdateQuery := `INSERT INTO v9.public.logs
+    (log_id, worker_id, component_id, execution_num, log_text, log_error, received_time)
+    VALUES ($1, $2, $3, $4, $5, $6, NOW())
+    ON CONFLICT (log_id) DO UPDATE SET log_text = $5, log_error = $6, received_time = NOW()`
+
+	_, err = driver.db.Exec(insertUpdateQuery, logID, workerID, compDBID, compLog.DedupNumber, logText, logError)
```
Comment (hjaensch7) [inline database/database.go]: again, what is not being taken from Exec?

File: database/database.go [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+}
+
+func (driver *Driver) PurgeDeploymentEntry(compID *worker.ComponentID) error {
+	compDBID, err := driver.FindComponentID(compID)
+	if err != nil {
+		return fmt.Errorf("could not get component ID to purge deploying entry: %w", err)
+	}
+
+	deleteQuery := `DELETE FROM v9.public.deploying WHERE component_id = $1`
+	_, err = driver.db.Exec(deleteQuery, compDBID, "initial_deployment")
```
Comment (hjaensch7) [inline database/database.go]: see above

File: pushHandler.go (hunk 5 above)
Comment (hjaensch7) [reply pushHandler.go:68]: issue #7

File: pushHandler.go [hunk 9]
```diff
@@ -59,54 +63,69 @@ func (h *pushHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+
+	// Setup the DB deploying entry
+	err = h.driver.EnterDeploymentEntry(&compID)
+	if err != nil {
+		log.Error.Println("Error starting deploy using db:", err)
+		return
```
Comment (hjaensch7) [inline pushHandler.go:74]: do we want to return here?

**Review by hjaensch7 — COMMENTED** [review]
Looks good, adding logging is useful and the worker struct is great. I only left a few comments where I had some confusion.

File: pushHandler.go (hunk 9 above)
Comment (Others) [reply pushHandler.go:74]: yes because it prevents double deploy

File: database/database.go (hunk 7 above)
Comment (Others) [reply database/database.go]: [check docs](https://golang.org/pkg/database/sql/#Result)

File: database/database.go (hunk 8 above)
Comment (Others) [reply database/database.go]: [check docs](https://golang.org/pkg/database/sql/#Result)

File: database/database.go (hunk 6 above)
Comment (Others) [reply database/database.go:94]: [check docs](https://golang.org/pkg/database/sql/#Result)

File: database/database.go (hunk 4 above)
Comment (Others) [reply database/database.go:163]: That's just for error reporting, the format doesn't really matter

**Review by hjaensch7 — DISMISSED** [review]
"Bit of a hack here" count = 300 but otherwise this refactor looks good. Will run tests on this branch.

File: database/database.go [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+	}
+
+	var deploymentStartTime string
+	upsertQuery := `INSERT INTO v9.public.deploying(component_id, deployment_reason)
+	VALUES ($1, $2) ON CONFLICT DO NOTHING
+	RETURNING to_char(deployment_start_time::timestamp at time zone 'UTC', 'YYYY-MM-DD"T"HH24:MI:SS"Z"')`
+	err = driver.db.QueryRow(upsertQuery, compDBID, "initial_deployment").Scan(&deploymentStartTime)
+	if err == nil {
+		return fmt.Errorf("there was a previous deployment starting at: %s", deploymentStartTime)
+	}
```
Comment (hjaensch7) [inline database/database.go]: Why is this throwing an error when err == nil?

**Review by hjaensch7 — APPROVED** [review]

**Review by pcodes — APPROVED** [review]
I have not read this and I refuse to. Looks good!

**Pull request merged**